h2o-core/src/main/java/water/fvec/Frame.java

```diff
@@ -1619,13 +1622,42 @@ public CSVStream(Chunk[] chks, String[] names, int nChunks, CSVStreamParams parm
       StringBuilder sb = new StringBuilder();
       if (names != null) {
         sb.append('"').append(names[0]).append('"');
-        for(int i = 1; i < names.length; i++)
+        for (int i = 1; i < names.length; i++)
           sb.append(_parms._separator).append('"').append(names[i]).append('"');
         sb.append('\n');
       }
       _line = StringUtils.bytesOf(sb);
       _chkRow = -1; // first process the header line
       _curChks = chks;
+      escapeCategoricalVecDomains(_curChks);
+    }
+
+    /**
+     * Escapes categorical levels of vectors and puts them in a map of escaped categorical levels.
+     * Only the domains with at least one level with an escaped quote are saved. If a domain does not need
+     * any escaping, it is considered better practice to reach to the `vec.domain()` method itself and not duplicate entries
+     * in memory here.
+     * @param chunks
+     */
+    private void escapeCategoricalVecDomains(final Chunk[] chunks) {
+
+      for (int i = 0; i < chunks.length; i++) {
+        final Vec vec = chunks[i].vec();
+        if (!vec.isCategorical()) continue;
+
+        final String[] originalDomain = vec.domain();
+        final String[] escapedDomain = new String[originalDomain.length];
+
+        boolean noEscapingRequired = true;
+        for (int level = 0; level < originalDomain.length; level++) {
+          escapedDomain[level] = escapeQuotesForCsv(originalDomain[level]);
+          noEscapingRequired = noEscapingRequired && escapedDomain[level].equals(originalDomain[level]);
+        }
+
+        if (!noEscapingRequired) {
+          escapedCategoricalLevels.put(i, escapedDomain);
```

Comment:
I wonder since this is indexed by column index if an simple  String[][] owuld not be faster and cheaper

Reply:
Cheaper in terms of memory - slightly. Faster ? Both are accessible in constant time, yet array would be faster. Tiny bit faster. But I don't know how to make it immutable inside, which I can do with a map. I'll use immutable map, WDYT ? Because of the immutability. The performance will be good anyway.

Reply:
yes, just thinking out loud, another option is immutable list, but since its sparse, map is probably best

Reply:
I like the String[][] suggestion, the issue with this approach is unnecessary autoboxing just to lookup the value

Reply:
As @michalkurka pointed out, there is implicit conversion from int to Integer involved. This makes me choose array over Map.

Reply:
I'll still keep the method and it will still return `String[][]` - to make it obvious the method sets the array of encoded categoricals in the constructor. One reference copy more.

Reply:
Done.